[MLIR][GPU][XeVM] Add XeVM target and XeVM dialect integration tests.

silee2 opened the pull request:
Covers remaining parts required for XeVM dialect integration testing. It has two high level components
- XeVM target and serialization support
- XeVM dialect integration tests using SYCL runtime

silee2 committed:
Add XeVM target and XeVM dialect integration tests.
Covers remaining parts required for XeVM dialect intgration testing.
It has two high level components
- XeVM target and serialization support
- XeVM dialect integration tests using SYCL runtime

silee2 commented:
FYI, @rengolin @adam-smnk @akroviakov @nbpatel

fabianmcg commented:
Nice, and congrats on getting access to LLVM!

I'll take a closer look later. But can we split this PR in one for the changes related to translation to LLVM IR and one for changes on serialization? In general it's easier to get reviews for smaller changes (that's not always possible, but I think in this case it is.)

adam-smnk reviewed (commented):
First pass with general comments.
I'll have to take a closer look but general direction is fine 👍

Perhaps, translation part could be moved to a separate PR but up to you.

silee2 commented:
> Nice, and congrats on getting access to LLVM!
> 
> I'll take a closer look later. But can we split this PR in one for the changes related to translation to LLVM IR and one for changes on serialization? In general it's easier to get reviews for smaller changes (that's not always possible, but I think in this case it is.)

Thanks for the suggestion. 
Will split into two PR.

silee2 commented:
> First pass with general comments. I'll have to take a closer look but general direction is fine 👍
> 
> Perhaps, translation part could be moved to a separate PR but up to you.

Will split the PR.

akroviakov reviewed (commented):
Looks good, but a few important points need to be addressed first: 

- removing SPIRV workaround for extensions 
- deciding what `gpu::CompilationTarget::Binary` means for XeVM.

silee2 committed:
Add XeVM to LLVMIR translation.

silee2 committed:
Address reviewer comments.

silee2 committed:
Merge branch 'addXeVMTranslate' into addXeVMTarget

Jianhui-Li commented on mlir/include/mlir/Target/LLVM/XeVM/Utils.h:
```diff
@@ -0,0 +1,39 @@
+  XeVMTargetAttr getTarget() const;
+
+protected:
+  XeVMTargetAttr target;
```
nit: consider target => xeTarget

silee2 replied:
Using `xeTarget` now.

Jianhui-Li commented on mlir/lib/Target/LLVM/XeVM/Target.cpp:
```diff
@@ -0,0 +1,257 @@
+    Operation &module, mlir::xevm::XeVMTargetAttr target,
+    const gpu::TargetOptions &targetOptions)
+    : ModuleToObject(module, target.getTriple(), "", {}, target.getO()),
+      target(target) {}
```
consideration for readability: target(target) -> xeTarget(target)

Jianhui-Li commented on mlir/lib/Target/LLVM/CMakeLists.txt:
```diff
@@ -209,3 +209,35 @@ if(MLIR_ENABLE_ROCM_CONVERSIONS)
+
+  LINK_LIBS PUBLIC
+  MLIRIR
+  MLIRExecutionEngineUtils
```
do we need to depend on ExeuctionEngine and GPUDialect?

silee2 replied:
XeVM target needs to access things like  gpu::TargetAttrInterface and gpu::TargetOptions

silee2 replied:
And base ModuleToObject requires makeOptimizingTransformer from ExecutionEngine

silee2 committed:
Merge remote-tracking branch 'origin/main' into addXeVMTarget

kurapov-peter commented on mlir/lib/Target/LLVM/XeVM/Target.cpp:
```diff
@@ -0,0 +1,257 @@
+    getOperation().emitError() << "SPIRV code size must be a multiple of 4.";
+    return std::nullopt;
+  }
+  StringRef bin(serializedSPIRVBinary->c_str(), serializedSPIRVBinary->size());
```
I think putting ocloc in would be too much of a hassle. SPIRV can be device-specific. We can leave it be.

akroviakov replied:
ocloc is bundled with [intel-compute-runtime](https://launchpad.net/ubuntu/+source/intel-compute-runtime) package, which, I suppose, is a common prerequisite for programming Intel GPUs. Nothing prevents us from returning binary SPIRV, while also supporting ocloc, just like nvidia allows binary ptx and cubin outputs. 
I'm not very familiar with the extent of SPIRV backend, can we meaningfully use `chip` and `O` in SPIRV backend the same way as in ocloc?

kurapov-peter commented on mlir/lib/Target/LLVM/XeVM/Target.cpp:
```diff
@@ -0,0 +1,257 @@
+  }
+  gpuMod.walk([&](LLVM::LLVMFuncOp funcOp) {
+    if (funcOp->hasAttr(gpu::GPUDialect::getKernelFuncAttrName())) {
+      funcOp.setIntelReqdSubGroupSize(16);
```
+1

silee2 committed:
Temp save.

silee2 committed:
Temp save.

silee2 committed:
Temp save.

silee2 committed:
Temp save.

silee2 committed:
Merge remote-tracking branch 'origin/main' into addXeVMTarget

silee2 committed:
Merge branch 'addXeVMTarget' into addXeVMTargetWIP

silee2 committed:
Refactor to prepare for future target triples.

silee2 committed:
Cleanup CMake file.

silee2 committed:
Use name xeTarget instead of target whenever applicable.

silee2 committed:
Update integration tests to use level zero runtime instead of SYCL runtime.

silee2 committed:
Add comments regarding binary format used for XeVM target.
SPIR-V binary is used as the binary format for now but
compileToBinary is already set up to call ocloc - the static compiler -
for creating Xe native binary.

chencha3 reviewed (approved):
LGTM

silee2 commented:
Split the translation part as a separate PR and merged #150696
This PR now consists of serialization and integration tests. 
Addressed all comments.
Some notable changes made:
- Refactored code and moved parts independent of target triple into `SerializeGPUModuleBase`. And subclass  `SpirSerializer` provides custom `init` and `moduleToObject` function.
- Added `compileToBinary` that used `ocloc` an offline compiler for SPIR-V to Xe native binary. 
- Integration tests are now using new level zero runtime wrapper.

@fabianmcg @Jianhui-Li @kurapov-peter @adam-smnk @adam-smnk Can you take another look at the PR?

adam-smnk commented on mlir/include/mlir/Target/LLVM/XeVM/Target.h:
```diff
@@ -0,0 +1,31 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_TARGET_XEVM_TARGET_H
```
nit: `MLIR_TARGET_LLVM_XEVM_TARGET_H`

silee2 replied:
Renamed.

adam-smnk commented on mlir/include/mlir/Target/LLVM/XeVM/Target.h:
```diff
@@ -0,0 +1,31 @@
+class MLIRContext;
+} // namespace mlir
+
+namespace mlir::xevm {
```
nit: `mlir::` can be removed

silee2 replied:
Removed.

adam-smnk commented on mlir/lib/Target/LLVM/XeVM/Target.cpp:
```diff
@@ -0,0 +1,417 @@
+  context.appendDialectRegistry(registry);
+}
+
+SerializeGPUModuleBase::SerializeGPUModuleBase(
```
Shouldn't `targetOptions`  be initialized here too?

adam-smnk reviewed (commented):
Overall looks good 👍
Left minor comments

silee2 committed:
Address more comments.

silee2 commented on mlir/include/mlir/Target/LLVM/XeVM/Utils.h:
```diff
@@ -0,0 +1,59 @@
+  SerializeGPUModuleBase(Operation &module, XeVMTargetAttr target,
+                         const gpu::TargetOptions &targetOptions = {});
+
+  XeVMTargetAttr getTarget() const;
```
Added docs line.

silee2 committed:
Break up long lines.

silee2 commented on mlir/test/Integration/Dialect/XeVM/GPU/xevm_block_dpas.mlir:
```diff
@@ -0,0 +1,135 @@
+      %base_pitch_a = arith.constant 32 : i32
+      %x = arith.constant 0 : i32
+      %y = arith.constant 0 : i32
+      %loaded_a = xevm.blockload2d %a, %base_width_a, %base_height_a, %base_pitch_a, %x, %y <{elem_size_in_bits=16 : i32, tile_width=16 : i32, tile_height=8 : i32, v_blocks=1 : i32, transpose=false, pack_register=false}> : (!llvm.ptr<1>, i32, i32, i32, i32, i32) -> vector<8xi16>
```
Added line breaks for all tests.

adam-smnk commented on mlir/lib/Target/LLVM/XeVM/Target.cpp:
```diff
@@ -0,0 +1,418 @@
+}
+
+namespace {
+class SpirSerializer : public SerializeGPUModuleBase {
```
nit: maybe `SPIRVSerializer` for consistency? Or does `Spir` refer to sth else?

silee2 committed:
Rename class.

silee2 committed:
Merge remote-tracking branch 'origin/main' into addXeVMTarget

silee2 committed:
Rename serializer class.

Kewen12 commented:
Hi @silee2, it seems this PR has build failure due to undefined reference. Could you please take a look? Thanks!
>undefined reference to `LLVMInitializeSPIRVTargetInfo'
